### PR TITLE
Merge charge deposition and maxwell solver routines into field interpolater/particle pusher branch.

### PR DESCRIPTION
**Pull request by pkim1818**
Merging in charge deposition and 1D/2D maxwell solvers.

### CODE REVIEW

File: src/espic/make_grid.py [hunk 1]
```diff
@@ -11,3 +11,17 @@ def __init__(self, num_points=100, x_min=-1, x_max=1):
...
+        self.num_points = num_points
+        self.x_min = x_min
+        self.x_max = x_max
+        self.y_min = y_min
+        self.y_max = y_max
+
+        self.xgrid = np.linspace(self.x_min, self.x_max, self.num_points)
+        self.ygrid = np.linspace(self.y_min, self.y_max, self.num_points)
+
+        self.grid = np.meshgrid(self.xgrid, self.ygrid)
```
Comment (leebr48) [inline src/espic/make_grid.py]:
```suggestion
        self.grid = np.meshgrid(self.xgrid, self.ygrid)
        self.size = self.grid.size
        self.shape = self.grid.shape
```

These changes simply add a couple of attributes to Uniform2DGrid that are in Uniform1DGrid. We could perhaps use inheritance to implement them, but I guess this is the simplest way.

File: src/espic/solve_maxwell.py [hunk 2]
```diff
@@ -34,3 +35,103 @@ def solve(self, rho):
         phi[1:-1] = solve_banded((1, 1), bands, rhs)
 
         return phi
+
+
+# Code taken from https://john-s-butler-dit.github.io/NumericalAnalysisBook/Chapter%2009%20-%20Elliptic%20Equations/903_Poisson%20Equation-Boundary.html
+# FIXME: for now, this assumes equal spacing in x and y.
+class MaxwellSolver2D:
```
Comment (leebr48) [inline src/espic/solve_maxwell.py:41]: Is there a way to make MaxwellSolver2D handle the 1D case easily? This would make the code simpler and easier to test and use. (I'm thinking of NumPy-style vectorization.)

File: src/espic/make_grid.py [hunk 3]
```diff
@@ -11,3 +11,17 @@ def __init__(self, num_points=100, x_min=-1, x_max=1):
...
+    def __init__(self, num_points=100, x_min=-1, x_max=1, y_min=-1, y_max=1):
+        self.num_points = num_points
+        self.x_min = x_min
+        self.x_max = x_max
+        self.y_min = y_min
+        self.y_max = y_max
+
+        self.xgrid = np.linspace(self.x_min, self.x_max, self.num_points)
+        self.ygrid = np.linspace(self.y_min, self.y_max, self.num_points)
+
```
Comment (leebr48) [inline src/espic/make_grid.py]:
```suggestion
        self.xgrid = np.linspace(self.x_min, self.x_max, self.num_points)
        self.ygrid = np.linspace(self.y_min, self.y_max, self.num_points)
```

File: src/espic/solve_maxwell.py [hunk 4]
```diff
@@ -34,3 +35,103 @@ def solve(self, rho):
...
+    def __init__(self, grid=Uniform2DGrid(), boundary_conditions=None):
+        self.grid = grid.grid
+        self.xgrid = grid.xgrid
+        self.ygrid = grid.ygrid
```
Comment (leebr48) [inline src/espic/solve_maxwell.py]: You might consider slightly changing the names of some variables to be consistent. For instance, `xgrid` -> `x_grid`, `ygrid` -> `y_grid`, `gridsize` -> `grid_size`, etc.

File: tests/maxwell_solver_test.py [hunk 5]
```diff
@@ -13,3 +15,109 @@ def test_1d_uniform():
...
+def test_2d_A():
+    ms = MaxwellSolver2D()
+    A = ms.set_A(4)
+
+    true_A = np.array([[-4, 1, 1, 0], [1, -4, 0, 1], [1, 0, -4, 1], [0, 1, 1, -4]])
+    np.testing.assert_allclose(A, true_A, atol=1e-9)
+
+
+def test_2d_boundary():
+    pass
```
Comment (leebr48) [inline tests/maxwell_solver_test.py:29]: What is this meant to do?

File: tests/maxwell_solver_test.py [hunk 6]
```diff
@@ -13,3 +15,109 @@ def test_1d_uniform():
...
+                / np.pi
+                * 1
+                / i
+                * np.cosh(i * np.pi * X / a)
+                / np.cosh(i * np.pi * b / a)
+                * np.sin(i * np.pi * Y / a)
+            )
+
+    # There are some issues that I believe are related to the convergence of an infinite Fourier series
+    # For now, these "magic" numbers should be kept fixed.
```
Comment (leebr48) [inline tests/maxwell_solver_test.py]:
```suggestion
    # FIXME There are some issues that I believe are related to the convergence of an infinite Fourier series
    # FIXME For now, these "magic" numbers should be kept fixed.
```
Just marking these lines so we can come back and address the issue later.

File: tests/maxwell_solver_test.py [hunk 7]
```diff
@@ -13,3 +15,109 @@ def test_1d_uniform():
...
+
+    err = np.abs(true_phi[1:-1, 1:-1] - phi[1:-1, 1:-1])
+    np.testing.assert_allclose(err, np.zeros(err.shape), atol=1e-2)
+
+
+# return phi, true_phi, X, Y
```
Comment (leebr48) [inline tests/maxwell_solver_test.py]:
```suggestion
```

File: tests/maxwell_solver_test.py [hunk 8]
```diff
@@ -13,3 +15,109 @@ def test_1d_uniform():
...
+                / np.pi
+                * 1
+                / i
+                * np.cosh(i * np.pi * X / a)
+                / np.cosh(i * np.pi * b / a)
+                * np.sin(i * np.pi * Y / a)
+            )
+    true_phi += -np.pi * (X**2 + Y**2)
+    # There are some issues that I believe are related to the convergence of an infinite Fourier series
+    # For now, these "magic" numbers should be kept fixed.
```
Comment (leebr48) [inline tests/maxwell_solver_test.py]:
```suggestion
    # FIXME There are some issues that I believe are related to the convergence of an infinite Fourier series
    # FIXME For now, these "magic" numbers should be kept fixed.
```

File: tests/maxwell_solver_test.py [hunk 9]
```diff
@@ -13,3 +15,109 @@ def test_1d_uniform():
...
+    err = np.abs(true_phi - phi)
+
+    # There are some issues here near the boundaries. The max error goes to about 0.22. Not sure what the problem is.
```
Comment (leebr48) [inline tests/maxwell_solver_test.py]:
```suggestion
    # FIXME There are some issues here near the boundaries. The max error goes to about 0.22. Not sure what the problem is.
```

File: tests/maxwell_solver_test.py [hunk 10]
```diff
@@ -13,3 +15,109 @@ def test_1d_uniform():
...
+    true_A = np.array([[-4, 1, 1, 0], [1, -4, 0, 1], [1, 0, -4, 1], [0, 1, 1, -4]])
+    np.testing.assert_allclose(A, true_A, atol=1e-9)
+
+
+def test_2d_boundary():
+    pass
+
+
+def test_2d_laplace():
+    """Test 2D Poisson solver for rho = 0. Ex taken from Ex 3.4 in Griffiths E&M, 4th ed"""
```
Comment (leebr48) [inline tests/maxwell_solver_test.py]:
```suggestion
    """Test 2D Poisson solver for rho = 0. Taken from Example 3.4 in Griffiths, "Introduction to Electrodynamics," 4th ed."""
```

**Review by leebr48 — COMMENTED** [review]
Very good overall. We can look at the boundary condition issues together soon.

File: src/espic/solve_maxwell.py (hunk 2 above)
Comment (pkim1818) [reply src/espic/solve_maxwell.py:41]: Not easily. The matrix in 1D is strictly tri-diagonal, whereas it's only block tri-diagonal in 2D. This affects the linear solve, since solving a tri-diagonal system is very fast, but less fast otherwise. I think it's okay to separate the two versions if one of them can use more specialized solvers.

A different possible fix would be having a generic MaxwellSolver class, which can switch to 1D or 2D.

File: src/espic/solve_maxwell.py (hunk 4 above)
Comment (pkim1818) [reply src/espic/solve_maxwell.py]: yeah I'll chase these down, thanks for the reminder.

File: tests/maxwell_solver_test.py (hunk 5 above)
Comment (pkim1818) [reply tests/maxwell_solver_test.py:29]: Ah right, I should have a test for setting the boundary conditions correctly.

File: src/espic/solve_maxwell.py (hunk 2 above)
Comment (pkim1818) [reply src/espic/solve_maxwell.py:41]: It's probably not impossible, but maybe not super important right now. I'd have to think of a slick way of doing it that doesn't make the code less readable.
